presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/JoinReorderingOptimizer.java

```diff
@@ -0,0 +1,130 @@
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.facebook.presto.statistics.PlanNodeStatistics;
+import com.facebook.presto.statistics.StatisticsCalculator;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class JoinReorderingOptimizer
+        implements PlanOptimizer
+{
+    private final StatisticsCalculator statisticsCalculator;
+
+    public JoinReorderingOptimizer(StatisticsCalculator statisticsCalculator)
+    {
+        this.statisticsCalculator = requireNonNull(statisticsCalculator, "statisticsCalculator can not be null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+
+        return SimplePlanRewriter.rewriteWith(new Rewriter(session), plan);
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final Session session;
+
+        public Rewriter(Session session)
+        {
+            this.session = session;
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
+        {
+            PlanNode leftRewritten = context.defaultRewrite(node.getLeft());
+            PlanNode rightRewritten = context.defaultRewrite(node.getRight());
+
+            PlanNodeStatistics leftStatistics = statisticsCalculator.calculateStatisticsForNode(session, leftRewritten);
+            PlanNodeStatistics rightStatistics = statisticsCalculator.calculateStatisticsForNode(session, rightRewritten);
+
+            boolean leftSizeKnown = !leftStatistics.getOutputSizeInBytes().isValueUnknown();
+            boolean rightSizeKnown = !rightStatistics.getOutputSizeInBytes().isValueUnknown();
+            double leftSize = leftStatistics.getOutputSizeInBytes().getValue();
+            double rightSize = rightStatistics.getOutputSizeInBytes().getValue();
+
+            boolean leftCountKnown = !leftStatistics.getOutputRowsCount().isValueUnknown();
+            boolean rightCountKnown = !rightStatistics.getOutputRowsCount().isValueUnknown();
+            double leftCount = leftStatistics.getOutputRowsCount().getValue();
+            double rightCount = rightStatistics.getOutputRowsCount().getValue();
+
+            boolean flipNeeded = (leftSizeKnown && rightSizeKnown && leftSize < rightSize)
```

Comment:
If one side of a join has a larger output size and the other has a larger row count, it will always get flipped.  We probably only want to check the row count if the size was not known.